Settings cleanup (#1050)
Akuli/porcupine: docs/settings.rst

.. |triangle| image:: ../porcupine/images/triangle.gif

Global and non-global settings
------------------------------

.. data:: global_settings

    You often need only this variable from the ``porcupine.settings`` module.
    It's a :class:`Settings` object that contains all settings that are the same everywhere in Porcupine.
    For example, it doesn't contain settings like "indentation is 4 spaces", which can be different in each tab.

.. class:: Settings

    Each instance of :class:`Settings` is basically a dict of option names and values,

>     but it does more than that;
>     it ensures that the values have the correct type,

Author: ```suggestion
    but it also ensures that the values have the correct type,
```

        shitty_bar = settings.get('bar', Optional[Path])
        reveal_type(shitty_bar)  # Any

    Use a type annotation to work around this (and make sure to write the
    same type two times)::

        good_bar: Path | None = settings.get('bar', Optional[Path])
        reveal_type(good_bar)  # Optional[Path]

    Before Python 3.10, you can't use the new ``|`` syntax as an argument to ``settings.get()``,
    even though it otherwise works with ``from __future__ import annotations``.
    The same goes for built-in generics,
    such as ``list[str]`` with lower-case ``list``.

    Options of mutable types are returned as copies, so things like
    ``settings.get('something', List[str])`` always return a new list.
    If you want to change a setting like that, you need to first get a copy
    of the current value, then modify the copy, and finally :meth:`set` it
    back. This is an easy way to make sure that change events run every
    time the value changes.

.. automethod:: Settings.debug_dump

Custom Options
--------------

Use :data:`global_settings` to store things that shouldn't reset every
time Porcupine is restarted.

.. automethod:: Settings.add_option

> Example::

Author: Example should probably be after a detailed description of the `add_option` method.

-------------------

Click *Porcupine Settings* in the *Edit* menu to open the dialog.

> It is meant to be used for editing :data:`global_settings`.

Author: ```suggestion
It allows the user to edit :data:`global_settings`.
```

A label that displays *text* is added to column 0 (see the ascii art above).
When setting the values, the converter passed to :func:`add_option` is used.

.. note::

    The *option_name* argument of the functions below
    must be an existing option on :data:`global_settings`.
    You can create it just before calling the functions below::

        from porcupine import settings
        from porcupine.settings import global_settings

        def validate_host(host: str) -> bool:
            ...

        ...

        global_settings.add_option('http_server_host', '127.0.0.1')

>         settings.add_entry('http_server_host', "HTTP server host:", validate_callback)

Author: ```suggestion
        settings.add_entry('http_server_host', "HTTP server host:", validate_host)
```
